servers/src/main/java/tachyon/master/file/FileSystemMaster.java

```diff
@@ -1478,8 +1517,182 @@ public void setState(long fileId, SetStateOptions options)
+        return filesToPersist;
+      }
+
+      Set<Long> scheduledFiles = mWorkerToAsyncPersistFiles.get(workerId);
+      for (long fileId : scheduledFiles) {
+        InodeFile inode = (InodeFile) mInodeTree.getInodeById(fileId);
+        if (inode.isCompleted()) {
+          fileIdsToPersist.add(fileId);
+          List<Long> blockIds = Lists.newArrayList();
+          for (FileBlockInfo fileBlockInfo : getFileBlockInfoList(fileId)) {
+            blockIds.add(fileBlockInfo.blockInfo.blockId);
+          }
+
+          filesToPersist.add(new PersistFile(fileId, blockIds));
+          // update the inode file persisence state
+          inode.setPersistenceState(PersistenceState.IN_PROGRESS);
+          scheduledFiles.remove(fileId);
+        }
+      }
+
+      mWorkerToAsyncPersistFiles.get(workerId).remove(fileIdsToPersist);
+
+      // write to journal
+      PersistFilesRequestEntry persistFilesRequest =
+          PersistFilesRequestEntry.newBuilder().addAllFileIds(fileIdsToPersist).build();
+      writeJournalEntry(
+          JournalEntry.newBuilder().setPersistFilesRequest(persistFilesRequest).build());
+      flushJournal();
+      return filesToPersist;
+    }
+  }
+
+  /**
+   * Updates a list of files as being persisted.
+   *
+   * @param fileIds the id of the files
+   * @throws FileDoesNotExistException when a file does not exist
+   */
+  private void setPersistingState(List<Long> fileIds) throws FileDoesNotExistException {
+    if (!fileIds.isEmpty()) {
+      LOG.info("Request file persistency: {}", fileIds);
+    }
+    for (long fileId : fileIds) {
+      InodeFile inode = (InodeFile) mInodeTree.getInodeById(fileId);
+      inode.setPersistenceState(PersistenceState.IN_PROGRESS);
+    }
+  }
+
+  /**
+   * Instructs a worker to persist the files.
+   *
+   * @param workerId the id of the worker that heartbeats
+   * @param persistedFiles the files that persisted on the worker
+   * @return the command for persisting the blocks of a file
+   * @throws FileDoesNotExistException if the file does not exist
+   * @throws InvalidPathException if the file path corresponding to the file id is invalid
+   */
+  public synchronized FileSystemCommand workerHeartbeat(long workerId, List<Long> persistedFiles)
       throws FileDoesNotExistException, InvalidPathException {
+    for (long fileId : persistedFiles) {
+      SetStateOptions.Builder builder = new SetStateOptions.Builder().setPersisted(true);
+      setState(fileId, builder.build());
+    }
+
+    // get the files for the given worker to checkpoint
+    List<PersistFile> filesToCheckpoint = pollFilesToCheckpoint(workerId);
+    if (!filesToCheckpoint.isEmpty()) {
+      LOG.info("Sent files {} to worker {} to persist", filesToCheckpoint, workerId);
+    }
+    FileSystemCommandOptions options = new FileSystemCommandOptions();
+    options.setPersistOptions(new PersistCommandOptions(filesToCheckpoint));
+    return new FileSystemCommand(CommandType.Persist, options);
```

Comment:
In fact, the second call will just invoke `create` from under FS, and upon different storage type, it may throw exception or overwrite the file. More handling of this should be addressed in a follow-up PR.